common/bytes_test.go

```diff
@@ -0,0 +1,261 @@
+	Count int
+	Data  string
+}
+
+func TestEncodeBytes(t *testing.T) {
+	assert, require := assert.New(t), require.New(t)
+
+	cases := []struct {
+		encoder  ByteEncoder
+		data     Bytes
+		expected string
+	}{
+		{HexEncoder, []byte{0x1a, 0x2b, 0x3c, 0x4d}, "1A2B3C4D"},
+		{B64Encoder, []byte("D!.3s"), "RCEuM3M="},
+		{RawB64Encoder, []byte("D!.3s"), "RCEuM3M"},
+	}
+
+	for i, tc := range cases {
+		clearByteEncoder()
+		SetByteEncoder(tc.encoder)
+		// encode the data
+		in := BData{Count: 15, Data: tc.data}
+		d, err := json.Marshal(in)
+		require.Nil(err, "%d: %+v", i, err)
+		// recover the data
+		out := BData{}
+		err = json.Unmarshal(d, &out)
+		require.Nil(err, "%d: %+v", i, err)
+		assert.Equal(in.Count, out.Count, "%d", i)
+		assert.Equal(in.Data, out.Data, "%d", i)
+		// check the encoding
+		view := BView{}
+		err = json.Unmarshal(d, &view)
+		require.Nil(err, "%d: %+v", i, err)
+		assert.Equal(tc.expected, view.Data)
+	}
+}
+
+/*** this is example code for the byte array ***/
+
+type Dings [5]byte
+
+func (d Dings) MarshalJSON() ([]byte, error) {
+	return GetByteEncoder().Marshal(d[:])
+}
+
+func (d *Dings) UnmarshalJSON(enc []byte) error {
+	var ref []byte
+	err := GetByteEncoder().Unmarshal(&ref, enc)
+	copy(d[:], ref)
+	return err
+}
+
+func TestByteArray(t *testing.T) {
+	assert, require := assert.New(t), require.New(t)
+
+	d := Dings{}
+	copy(d[:], []byte("D!.3s"))
+
+	cases := []struct {
+		encoder  ByteEncoder
+		data     Dings
+		expected string
+	}{
+		{HexEncoder, Dings{0x1a, 0x2b, 0x3c, 0x4d, 0x5e}, "1A2B3C4D5E"},
+		{B64Encoder, d, "RCEuM3M="},
+		{RawB64Encoder, d, "RCEuM3M"},
+	}
+
+	for i, tc := range cases {
+		clearByteEncoder()
+		SetByteEncoder(tc.encoder)
+		// encode the data
+		d, err := json.Marshal(tc.data)
+		require.Nil(err, "%d: %+v", i, err)
+		// recover the data
+		out := Dings{}
+		err = json.Unmarshal(d, &out)
+		require.Nil(err, "%d: %+v", i, err)
+		assert.Equal(tc.data, out, "%d", i)
+		// check the encoding
+		view := ""
+		err = json.Unmarshal(d, &view)
+		require.Nil(err, "%d: %+v", i, err)
+		assert.Equal(tc.expected, view)
+	}
+
+	// Test invalid data
+	invalid := []byte(`"food"`)
+	clearByteEncoder()
+	SetByteEncoder(HexEncoder)
+	ding := Dings{1, 2, 3, 4, 5}
+	parsed := ding
+	require.Equal(ding, parsed)
+	// on a failed parsing, we don't overwrite any data
+	err := json.Unmarshal(invalid, &parsed)
+	require.NotNil(err)
+	assert.Equal(ding, parsed)
+}
+
+// I guess this is "proper"?  Seems funny though...
+// echo -n "deadbQ==" | base64 -D | od -H
+// echo -n "deadb9==" | base64 -D | od -H
+// both ---> 6f9de675
+// func TestWTF(t *testing.T) {
+// 	assert := assert.New(t)
+// 	url := base64.URLEncoding
+// 	std := base64.StdEncoding
+
+// 	cases := []struct {
+// 		s string
+// 		e *base64.Encoding
+// 	}{
+// 		// these fail...
+// 		{"D4_a--1=", url},
+// 		{"D4_a0-1=", url},
+// 		{"D4_a__1=", url},
+// 		{"D4/a++1=", std},
+// 		{"D4_a0-9=", url},
+// 		{"Finey-1=", url},
```

Comment:
For a second I thought this was an ode to Hal Finney but then noticed it was just one `n` :)